Correct formatting of preview widget fields

droidmonkey opened the pull request:
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change which fixes an issue)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )
* Fix #3701 - replace QLabel with QTextEdit to enable scrolling of notes

* Notes are plain text. They will remain as plain text and hyperlinks will not be enabled in the notes. Until the notes editor is moved to a rich text / html editor this will remain the case.

* Convert username and password fields in preview pane to QLineEdit's to allow for full copying and viewing if larger than the field width.

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )
![image](https://user-images.githubusercontent.com/2809491/68530536-cdf71380-02d6-11ea-9f71-e89008a93ee0.png)

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Tested on Windows

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation, and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

droidmonkey commented:
ping @0pLuS0 and @ei-ke and @MyXelf

0pLuS0 commented:
@droidmonkey HI

If this is for the ```Notes```, hmm since I used 2.4.3, I've personally gotten use to the idea of it having it's own Tab, not sure why it was thought it was a bad idea having it's own tab and now removing it...

Hmm

I think it's best to have the Notes Tab put back in...

droidmonkey commented:
This is how it is now, we aren't going back. Attributes and Attachments were combined into the Advanced tab.

0pLuS0 commented:
@droidmonkey ok thanks, I appreciate your time too by the way, keep up the great work!

ei-ke commented:
@droidmonkey Thank you for the quick fix. Just build your pull request on linux without issues. For the Borders @MyXelf metioned I could live with how it is now. But maybe the background color of the notes field could be kept the save as the rest of the UI?
I've tried this with QT Designer and somehow got it working, but honestly I believe it's better if you do it on your own :)

MyXelf commented:
As visual aids, here are the screenshots of the 3 variants in Linux (all showing the scrollbars).

Widget with Default Background and StyledPanel (this PR as is):
![keepassxc---#3727---preview-notes---01-bgblack-wborders](https://user-images.githubusercontent.com/4689660/67915338-11c87b00-fb69-11e9-808c-73b93bb3fa12.png)

Widget with Transparent Background and Border:
![keepassxc---#3727---preview-notes---02-nobg-wborders](https://user-images.githubusercontent.com/4689660/67915841-5ef91c80-fb6a-11e9-94d8-174c7f3b6755.png)

Widget with Transparent Background and No Border (my terrible idea) ;)
![keepassxc---#3727---preview-notes---03-nobg-noborders](https://user-images.githubusercontent.com/4689660/67915876-73d5b000-fb6a-11e9-9a5d-86d9268877ef.png)

At the end the visual appearance is OS-specific and user-specific. The light vs dark theme is another issue (that shouldn't interfere, but now it does). And the default look of the `QPlainTextEdit` is definitely suggesting "input", which is not the case.

If requested I could build the screenshots with a light theme.

ei-ke commented:
@MyXelf now I'd also like the third option: transparent background, no border :)

droidmonkey commented:
You done convinced me too :sweat_smile:

MyXelf commented:
:rofl:

The other option to avoid the floating scrollbar is to make the widget adjust to the content of the Notes field and let the "container" do the scrolling if needed. This would be a better way. But right now the Entry Preview is needing a lot of consistency fixes.

I'm about to provide the light theme screenshots.

MyXelf commented:
Here are the screenshots of the 3 variants in Linux using light theme Breeze (all showing the scrollbars).

Widget with Default Background and StyledPanel (this PR as is):
![keepassxc---#3727---preview-notes---light---01-bgwhite-wborders](https://user-images.githubusercontent.com/4689660/67950373-d0af8580-fbbf-11e9-8f02-33b1cff200ad.png)

Widget with Transparent Background and Border:
![keepassxc---#3727---preview-notes---light---02-nobg-wborders](https://user-images.githubusercontent.com/4689660/67950388-d907c080-fbbf-11e9-862f-e61ddb1e76d6.png)

Widget with Transparent Background and No Border:
![keepassxc---#3727---preview-notes---light---03-nobg-noborders](https://user-images.githubusercontent.com/4689660/67950423-e91fa000-fbbf-11e9-9129-1f317af14eb7.png)

HTH

droidmonkey commented:
Hmmm I actually like the idea of scrolling the whole preview widget instead. It makes more sense to me when the border is removed

MyXelf commented:
Is the `QWidget` able to scroll the content? Or we would need an additional container?

Unless you are able to make the changes in no time it is not worth the effort (IMHO). What's another drop in the ocean? I would focus on the functionality and leave the UX for later (I want to get there eventually). As I said there is a lot that must be done in the Preview UI (misaligned controls, toggle buttons, lot of spacers, inconsistencies between the tabs, to name a few).

droidmonkey commented:
You would replace the qwidget with a qscrollwidget

0pLuS0 commented:
@droidmonkey are the Notes always going to visible or will there be some way to hide them when not needed?

droidmonkey commented:
There is a setting to hide notes by default

0pLuS0 commented:
@droidmonkey ok thanks, keep up the good work! :+1:

phoerious commented on src/gui/EntryPreviewWidget.cpp:
```diff
@@ -189,28 +189,31 @@ void EntryPreviewWidget::setPasswordVisible(bool state)
-        notesLabel->setText(hyperlinkNotes.replace(QRegExp("(\\w+:\\/\\/\\S+)"), "<a href=\"\\1\">\\1</a>"));
-        notesLabel->setTextInteractionFlags(Qt::TextBrowserInteraction);
+        hyperlinkNotes.replace("\n", "<br/>");
+        notesWidget->appendHtml(hyperlinkNotes.replace(QRegExp("(\\w+:\\/\\/\\S+)"), "<a href=\"\\1\">\\1</a>"));
```
hyperlinkNotes needs proper HTML escaping before doing any replacements and then appending it as HTML to the widget.

droidmonkey replied:
Good point

droidmonkey replied:
I also need to fix the styling of this

0pLuS0 commented:
Well the DEV(s) are at this PLEASE make the ```Password Preview``` show the entire password, which it doesn't on passwords with long characters...

Keep up the great work!

THANKS

droidmonkey commented:
Will do

0pLuS0 commented:
> Will do

:+1:

droidmonkey commented:
This is now final and addresses all concerns. One caveat is that the hyperlinks in the notes have been disabled. I spent a good 3 hours trying to satisfy all criteria for our notes (allow monospace font, obey whitespace, etc. etc.) and adding hyperlinks not only adds a ton of code, but it also introduces several edge cases due to parsing of the notes.

To this end, when notes become a rich text / html editor we can bring hyperlinks back into the fold.
